app/javascript/src/components/Editor.tsx

```diff
@@ -34,7 +34,7 @@ interface EditorProps {
 class Editor extends React.Component<EditorProps, EditorState> {
   engine: DiagramEngine
   model: DiagramModel
-  lastSavedState: ApplicationState
+  lastSavedState: string
```

Comment:
Can you explain why this change is necessary?  The issue I have with this, is that the string type is weaker than the ApplicationState type.  Now pretty much any object could be serialized into a string, and cause type errors down the line.

Reply:
This came up in a discussion with @efatsi. When it was an application state, it was causing problems with line 325, since `lastSavedState` would sometimes update at the same time as `newState`. He could give a little more insight into exactly why, since he had worked on a potential fix in 0899f7ea. He mentioned the above approach would work, since a string for `lastSavedState` would not cause that strange problem of `lastSavedState` updating when `newState` was updated, causing line 326 of the old version to report that there was no change when there was, resulting in a save not happening where it should. 

I understand from your explanation why this may cause errors down the line; do you think we should look into Eli's commit to try to get something like that to work?

Reply:
Yeah, portions of the app don't work idempotently, so updates to the current state pollute the `lastSavedState`. My commit should do the trick if we want to keep `lastSavedState` from being a string (or at least a subset of it that dealt with this). It writes a clean dup of the state to `lastSavedStaet`. It's sitting in a branch still so that's why this is still a problem on master.

That said, I'm not all that worried about the `lastSavedState` being a string since as it stands now it's just being used for this one purpose (and in this purpose we turn it into a string for the equality check). But I don't have a brain for Typescript yet so differ to others for best practices here.